modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointMarkersStorage.java

```diff
@@ -0,0 +1,210 @@
+        }
+    }
+
+    /**
+     * Callback at the start of the checkpoint.
+     *
+     * <p>Creates a start marker for a checkpoint.
+     *
+     * @param checkpointId Checkpoint id.
+     */
+    public void onCheckpointBegin(UUID checkpointId) throws IgniteInternalCheckedException {
+        assert !checkpointIds.contains(checkpointId) : checkpointId;
+
+        Path checkpointStartMarker = checkpointDir.resolve(checkpointMarkerFileName(checkpointId, CHECKPOINT_START_MARKER));
+
+        try {
+            createFile(checkpointStartMarker);
+        } catch (IOException e) {
+            throw new IgniteInternalCheckedException("Could not create start checkpoint marker: " + checkpointStartMarker, e);
+        }
+
+        checkpointIds.add(checkpointId);
+    }
+
+    /**
+     * Callback at the end of the checkpoint.
+     *
+     * <p>Creates an end marker for a checkpoint.
+     *
+     * @param checkpointId Checkpoint id.
+     */
+    public void onCheckpointEnd(UUID checkpointId) throws IgniteInternalCheckedException {
+        assert checkpointIds.contains(checkpointId) : checkpointId;
+
+        Path checkpointEndMarker = checkpointDir.resolve(checkpointMarkerFileName(checkpointId, CHECKPOINT_END_MARKER));
+
+        try {
+            createFile(checkpointEndMarker);
+        } catch (IOException e) {
+            throw new IgniteInternalCheckedException("Could not create end checkpoint marker: " + checkpointEndMarker, e);
+        }
+
+        if (checkpointIds.size() >= earliestCheckpointChangesThreshold) {
+            for (Iterator<UUID> it = checkpointIds.iterator(); it.hasNext(); ) {
+                UUID id = it.next();
+
+                if (!id.equals(checkpointId)) {
+                    removeCheckpointMarkers(id);
+
+                    it.remove();
+                }
+            }
+        }
+    }
+
+    private void removeCheckpointMarkers(UUID checkpointId) {
+        Path startMarker = checkpointDir.resolve(checkpointMarkerFileName(checkpointId, CHECKPOINT_START_MARKER));
+        Path endMarker = checkpointDir.resolve(checkpointMarkerFileName(checkpointId, CHECKPOINT_END_MARKER));
+
+        if (exists(startMarker)) {
+            startMarker.toFile().delete();
+        }
+
+        if (exists(endMarker)) {
+            endMarker.toFile().delete();
+        }
+    }
+
+    /**
+     * Checks that the directory contains only paired (start and end) checkpoint markers.
+     */
+    private static void checkCheckpointDir(Path checkpointDir) throws IgniteInternalCheckedException {
+        assert isDirectory(checkpointDir) : checkpointDir;
+
+        try {
+            Map<Boolean, List<Path>> files = list(checkpointDir)
+                    .collect(partitioningBy(path -> parseCheckpointIdFromMarkerFile(path) != null));
+
+            if (!files.get(false).isEmpty()) {
+                throw new IgniteInternalCheckedException(
```

Comment:
At the moment, no temporary files are expected and there will be an error, later this will be implemented differently. 
Added test: `org.apache.ignite.internal.pagememory.persistence.checkpoint.CheckpointMarkersStorageTest#testTmpMarkersFiles`